opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74023
* #70897
* #70665
* #68101
* __->__ #58348
* #55864

Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

mingfeima committed:
opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"




[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29463782).

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29463782).

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29463782).

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

jbschlosser commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2999,6 +2999,18 @@
   dispatch:
     CompositeExplicitAutograd: mkldnn_convolution_backward
 
+- func: mkldnn_convolution_transpose(Tensor self, Tensor weight, Tensor? bias, int[] padding, int[] output_padding, int[] stride, int[] dilation, int groups) -> Tensor
+  dispatch:
+    CompositeExplicitAutograd: mkldnn_convolution_transpose
+
+- func: mkldnn_convolution_transpose_backward_input(int[] self_size, Tensor grad_output, Tensor weight, int[] padding, int[] output_padding, int[] stride, int[] dilation, int groups, bool bias_defined) -> Tensor
+
+- func: mkldnn_convolution_transpose_backward_weights(int[] weight_size, Tensor grad_output, Tensor self, int[] padding, int[] output_padding, int[] stride, int[] dilation, int groups, bool bias_defined) -> (Tensor, Tensor)
+
+- func: mkldnn_convolution_transpose_backward(Tensor self, Tensor grad_output, Tensor weight, int[] padding, int[] output_padding, int[] stride, int[] dilation, int groups, bool[3] output_mask) -> (Tensor, Tensor, Tensor)
+  dispatch:
+    CompositeExplicitAutograd: mkldnn_convolution_transpose_backward
```
FYI I've been working to collapse the various convolution ops into a single general set of `convolution` + `convolution_backward` that handle routing to the correct backends.

Towards that effort, is it possible to demote the ops here to helper functions that are not exposed in `native_functions.yaml`?

mingfeima replied:
sure, i noticed your related PRs. Feel free to go ahead, i will rebase accordingly :)

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29463782).

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima commented:
TODO:
- [ ]: update backward part to match https://github.com/pytorch/pytorch/commit/e6c3aa388050f774df3f004d8c0917a0dfbf97d6

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29463782).

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3101,6 +3101,10 @@
 
+- func: mkldnn_convolution_transpose(Tensor self, Tensor weight, Tensor? bias, int[] padding, int[] output_padding, int[] stride, int[] dilation, int groups) -> Tensor
+  dispatch:
+    CompositeExplicitAutograd: mkldnn_convolution_transpose
```
Now that we have a single `convolution` and `convolution_backward` op, maybe these mkldnn variants should be folded directly into `convolution`? (cc @albanD, I'm not super familiar with how the new implementation works)

Then again it looks like we already have an existing `mkldnn_convolution` op that wasn't folded into `convolution` yet.

albanD replied:
Yes they should be folded in there and shouldn't have an entry in native_functions.yaml anymore and the general convolution will call into that directly.

mingfeima replied:
at the current moment, `mkldnn_convolution_transpose` is treated in the same way as other `mkldnn_` functions and `cudnn`/`miopen` counterpart, feel free if additional changes are required.

mingfeima replied:
this entry has removed and used stub instead.

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -2542,6 +2542,9 @@
 - name: mkldnn_convolution(Tensor self, Tensor weight, Tensor? bias, int[] padding, int[] stride, int[] dilation, int groups) -> Tensor
   self, weight, bias: "grad.defined() ? convolution_backward(grad, self, weight, bias->sizes(), stride, padding, dilation, /*transposed=*/ false, /*output_padding=*/ std::vector<int64_t>(padding.size(), 0), groups, grad_input_mask) : std::tuple<Tensor, Tensor, Tensor>()"
 
+- name: mkldnn_convolution_transpose(Tensor self, Tensor weight, Tensor? bias, int[] padding, int[] output_padding, int[] stride, int[] dilation, int groups) -> Tensor
```
This is not needed, the autograd is already handled at the `convolution()` op level

mingfeima replied:
this entry has been removed.

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -11211,6 +11211,7 @@ def ref_pairwise_distance(input1, input2):
            aten_name='conv_transpose1d',
            aliases=('conv_transpose1d',),
            dtypes=floating_types_and(torch.int64),
+           dtypesIfCPU=floating_types_and(torch.int64, torch.bfloat16),
```
Why are you restricting the test dtypes here?

mingfeima replied:
conv_transpoed2d is enabled with bfloat16 with this pull request so that conv_transposed1d is enabled with bfloat16 ' as well since 3d input is unsqueezed to 4d in Convolution.cpp

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3101,6 +3101,10 @@
   dispatch:
     CompositeExplicitAutograd: mkldnn_convolution
 
+- func: mkldnn_convolution_transpose(Tensor self, Tensor weight, Tensor? bias, int[] padding, int[] output_padding, int[] stride, int[] dilation, int groups) -> Tensor
```
This entry is not needed. The others functions are still here for BC reasons and will be removed as well soon.
We expect to only have `convolution` and `convolution_backward` here in the end state.

mingfeima replied:
both the foward and backward of mkldnn_convolution_transpose entries have been removed from `native_functions.yaml`, use stub instead.

albanD reviewed (changes requested):
The test change don't look right?

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -1366,6 +1374,21 @@ at::Tensor _convolution(
+        weight = weight.contiguous(backend_memory_format);
+        bias = bias.defined() ? bias.contiguous() : bias;
+      }
+      output = at::mkldnn_convolution_transpose(
```
both the foward and backward of mkldnn_convolution_transpose entries have been removed from `native_functions.yaml`, use stub instead.

albanD commented on torch/overrides.py:
```diff
@@ -149,6 +149,7 @@ def get_ignored_functions() -> Set[Callable]:
         torch.logspace,
         torch.mkldnn_adaptive_avg_pool2d,
         torch.mkldnn_convolution,
+        torch.mkldnn_convolution_transpose,
```
This is not needed anymore right?

mingfeima replied:
Fixed!

albanD reviewed (commented):
native function change looks good.
Letting @VitalyFedyunin finish the merge.

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

mingfeima committed:
Update on "opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU"


Differential Revision: [D29463782](https://our.internmc.facebook.com/intern/diff/D29463782)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29463782).

malfet commented:
I had to revert the change as it broke dozens of OSS CI tests, see https://hud.pytorch.org/pytorch/pytorch/commit/479e0d64e607d48fc61a958caf6aeaf165e3d45d

(And looks very similar to previous revert)
